### DISCUSSION TITLE
How to use LinkedIn for OpenID

### DISCUSSION THREAD

**Discussion by chrstphrrbn** [pocketbase/pocketbase#3799] [Q&A]
I set up the OpenID auth provider as per https://learn.microsoft.com/en-us/linkedin/consumer/integrations/self-serve/sign-in-with-linkedin-v2 to use LinkedIn as provider.

After logging in at LinkedIn I get a 400 on the POST to /api/collections/users/auth-with-oauth2 with the message
`{
  "errorMessage": "oauth2: \"invalid_client\" \"Client authentication failed\""
}`

I tried running Pocketbase in debug mode but I get no additional information. Does anyone have any idea how to troubleshoot this or even what the error message actually means?

**Comment by ganigeorgiev** [comment]
Double check that you use the correct client id and secret and that there are no typos or spaces.

For the Linkedin API error references you can check https://learn.microsoft.com/en-us/linkedin/shared/authentication/client-credentials-flow#api-error-details.

**Comment by chrstphrrbn** [comment]
They work if I use them in Postman.

Is there any way to trace the calls made from PocketBase?

**Reply by ganigeorgiev** [reply]
You can try inspecting your network traffic with [Wireshark](https://www.wireshark.org/) or other similar software and compare the sent headers with your Postman call but I remember testing the OIDC client with Authentik and it should work correctly.

Alternatively, you can try [building PocketBase locally](https://github.com/pocketbase/pocketbase#running-and-building) and debug the source calls in https://github.com/pocketbase/pocketbase/blob/985ab1e5b79a328b516decca3d1d7dc8edfcbba0/forms/record_oauth2_login.go#L127-L152
(_I believe the last `provider.FetchToken` call is failing for some reason in your case_).

**Comment by hungcrush** [comment]
Hi @ganigeorgiev I got same error message

`2023/11/22 06:55:43 oauth2: "invalid_client" "Client authentication failed"`

**Reply by ganigeorgiev** [reply]
I currently don't have a LinkedIn account to test it, but will try sometime later this week/weekend to create one and will try the OIDC integration with it myself.

I remember that I've tested the OIDC implementation with Authentik and it worked fine. Other users (see recently https://github.com/pocketbase/pocketbase/discussions/3787) also confirmed that it works fine for them for example with Keycloak.

It is possible that LinkedIn has some additional specific requirements for the integration, eg. some custom header that we'll have to send with the request. If that's the case then a dedicated LinkedIn provider may be needed.

**Reply by hungcrush** [reply]
Yeahhh Hi @ganigeorgiev, after debug many many times. I tested with postman, and I realize that we just need to send without code_verifier.

After I tried to remove `oauth2.SetAuthURLParam("code_verifier", form.CodeVerifier),` it works well... 


maybe it only happened with Linkedin ?

**Reply by ganigeorgiev** [reply]
Hm, it is possible if they don't support the PKCE flow and doesn't allow extra query parameters for some reason.

To test this, could you try defining any other custom query parameter, eg. `abc=1` in your Postman call to see if it will trigger the same error?

**Reply by hungcrush** [reply]
@ganigeorgiev it works well.

**Reply by ganigeorgiev** [reply]
Hm, then something else is going on. Are you sure that you are sending the correct `code_verifier` and not some random string? Note that the `code_verifier` value is linked to the initial `code_challenge` available with the signin url from the `listAuthMethods()` call (or if you use the "all-in-one" flow the url can also be inspected using the `urlCallback` config option).

**Reply by hungcrush** [reply]
Yes. I'm sure that I sending correct `code_verifier`.
![CleanShot 2023-11-22 at 15 52 50@2x](https://github.com/pocketbase/pocketbase/assets/8098916/51cae713-8760-49e6-8344-6ad7b4b99f9a)


I tested another case, send `code_verifier` with empty value, it works well..

**Reply by ganigeorgiev** [reply]
This is really confusing. Again, please note that the aboce `codeVerifier` is linked to the codeChallenge available with the signin url. It is expected to work **only** if the signin url has the related `code_challenge` param.

**Reply by ganigeorgiev** [reply]
Hm, based on this documentation https://learn.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow-native it seems that PKCE is supported but there is a separate authtorization endpoint for it
```
https://www.linkedin.com/oauth/native-pkce/authorization
```

Try to replace the Auth URL in the OIDC settings with the above url and check if it will work.

**Reply by hungcrush** [reply]
@ganigeorgiev It not works.

![CleanShot 2023-11-22 at 16 12 35@2x](https://github.com/pocketbase/pocketbase/assets/8098916/acbed934-c8d3-4764-9cae-439341b2e13b)

**Reply by ganigeorgiev** [reply]
Another thing also mentioned in the above documentation it seems that you may have to contact Linkedin to enable the PKCE flow. I have no idea why this need to be enabled on per user basis and not available globally.

> Once you have the app created, please reach out to your point of contact at LinkedIn, and we will enable PKCE OAuth 2 flow for your app.

**Reply by ganigeorgiev** [reply]
@hungcrush See the above comment.

**Reply by hungcrush** [reply]
I got it. Thank you for your patient support

**Comment by ganigeorgiev** [comment]
I've tried to create a Linkedin account but I don't think I'll be able to test it (at least not today). It seems that in order to create a LinkedIn OAuth2 app you need first to create a public "company page", but in order to create that page you also seems to need at least 1 another contact and to wait at least 24h because my account was new (this is based on the error messages I got so far).

I'm not sure that I'm reading the LinkedIn documentation correctly and I don't want to dedicate more time for this at the moment as I'm already behind my schedule for today.

To summarize - PocketBase by default attempts to use the Code + PKCE flow. So far most providers that don't support the PKCE flow usually just ignore the extra query parameters. This unfortunately doesn't seem to be the case with LinkedIn.

Additionally based on https://learn.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow-native#redirecting-the-user, LinkedIn seems to use a separate auth endpoint for the PKCE flow.

The problems with the above are:
- You may need to contact LinkedIn to enable it for you (_I don't understand why this is needed, maybe it is in beta?_).
   This assumption is based on the following docs note:
   > Once you have the app created, **please reach out to your point of contact at LinkedIn**, and we will enable PKCE OAuth 2 flow for your app.

- Even if the above is enabled I'm still still not sure actually whether it will work as based on the @hungcrush [screenshot from earlier](https://github.com/pocketbase/pocketbase/discussions/3799#discussioncomment-7639859) PKCE may not be supported for their newer OIDC apis (note that the [OIDC docs](https://learn.microsoft.com/en-us/linkedin/consumer/integrations/self-serve/sign-in-with-linkedin-v2) linked by @chrstphrrbn uses `/oauth/v2/...` endpoints while the PKCE endpoint in the docs is just `/oauth/...`).


I'm not sure that I'll be able to help further with troubleshooting the LinkedIn OIDC integration, but what we can do make the `CodeVerifier` form field optional and skip adding the `code_verifier` query parameter to the token request if it is empty to accommodate providers that have more strict query parameters requirement. I'll consider doing this for the next release together with some other minor planned additions to the OAuth2 integration.

**Reply by chrstphrrbn** [reply]
I'm not familiar with PKCE but FYI I've contacted LinkedIn to enable PKCE to try the end point mentioned above. Will let you know when that's done!

**Reply by ganigeorgiev** [reply]
@hungcrush, @chrstphrrbn I've pushed in the `develop` branch option to toggle the PKCE for the OIDC providers from the Admin UI settings. The change will be part of the next v0.20.0 release once finalized (there is no ETAs yet). If you don't want to wait, you can try building the `develop` on your own:

0. [Install Go 1.21+](https://go.dev/doc/install)
1. Checkout or clone the `develop` branch
   ```sh
   git clone https://github.com/pocketbase/pocketbase.git --branch develop --single-branch
   ```
2. Navigate to `examples/base`
3. Run `GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build` (_https://go.dev/doc/install/source#environment_)
4. Start the created executable by running `./base serve`.

**Reply by hungcrush** [reply]
@ganigeorgiev Thank you so much

**Reply by hungcrush** [reply]
@ganigeorgiev It works very well.. I have a question,

I usually use `--debug` when I am developing hooks (JS). However, it seems you have removed isDebug(), so now how can I debug in production?

**Reply by ganigeorgiev** [reply]
@hungcrush Yes, with the next release and due to the Go 1.21 `slog` integration, the flag is no longer needed and instead the debug logs will be shown in the refactored _Admin UI > Logs_ section.

Note that to enable the debug logs you'll need to define the "Min log level" from the "Logs settings" popup and set it to `-4` or lower.

All currently used log levels are:
```
-4: DEBUG
0:  INFO
4:  WARN
8:  ERROR
```

Also note that as of now the PocketBase slog handler is implemented to write the logs with debounce and on batches (aka. there is a minor tradeoff between performance and durability):

- 3 seconds after the last debounced log write
- when the batch threshold, currently 200, is reached (this is primarily to prevent the memory to grow unconstrained)
- right before app termination to attempt saving whatever is in the existing logs queue

**Reply by chrstphrrbn** [reply]
@hungcrush Did you get it to work with LinkedIn?

I installed 0.20 and unchecked PKCE, but I still get the same error message.

If it works for you, could you be kind enough to share your settings for the OIDC provider in Pocketbase?

**Reply by hungcrush** [reply]
@chrstphrrbn It works very well after I upgraded to 0.20 with unchecked PKCE. You can refer my config:
![CleanShot 2023-12-12 at 16 08 21@2x](https://github.com/pocketbase/pocketbase/assets/8098916/2e03c3ce-bedb-4339-b3fc-cee33edfb534)

**Reply by chrstphrrbn** [reply]
Thank you @hungcrush , that got me one step closer!

Now the logs get this instead of creating a new user:
```
failed to fetch OAuth2 user profile via https://api.linkedin.com/v2/userinfo (500):
{"status":500,"code":"GATEWAY_INTERNAL_ERROR","message":"INTERNAL_SERVER_ERROR"}
```

Did the logging changes you wrote about above by any chance allow me to somehow log the request to see what is being sent?

**Reply by chrstphrrbn** [reply]
Problem solved! The app on LinkedIn needs to be approved by a company page to allow that request.

**Reply by ElvinKyungu** [reply]
![image](https://github.com/pocketbase/pocketbase/assets/85198945/ac0f18d3-e35c-453f-a60a-f6c2bb6d2ca3)

I'm having trouble putting the redirect link in linkedin, first I've already added the connection with Google and I tried to put the same redirect link that is in Google also for OCID with Linkedin but alas, it doesn't work.
They say it has to be different!
@ganigeorgiev @chrstphrrbn

**Comment by Anas-Dew** [comment]
I still face invalid_client error

**Reply by ganigeorgiev** [reply]
Sorry but I don't think anyone will be able to assist you without details what steps did you follow and whether you've checked the past comments and configuration suggestions (make sure that the PKCE flag is disabled).

To help you with the debugging, you can also start the executable with `--dev` and/or check the _Dashboard > Logs_ for more detailed error logs.

**Reply by Anas-Dew** [reply]
I am using nodejs as backend and how to disable PKCE flag?

**Reply by ganigeorgiev** [reply]
Please read the previous comments (_and more specifically the "Support PKCE" option_). Also _"using nodejs as backend"_ doesn't really help much and I don't see how this is related to PocketBase.
Additionally, if you haven't already please read https://pocketbase.io/docs/how-to-use/.

If you need further help, feel free to open a new Q&A discussion, but please try to elaborate more on your setup with clear steps to reproduce your issue otherwise it is just lazy and low effort.
